neusomatic/python/sequencing_features.py

```diff
@@ -0,0 +1,379 @@
+        delta_q = length if (is_aligned or op == cigar_insertion) else 0
+        while current_i < len(ref_pos_s):
+            diff = ref_pos_s[current_i] - pos_r
+            if diff < delta_r:
+                output[ref_pos_s[current_i]]=[count + diff, (pos_q + diff) if delta_q else None]
+                current_i +=1
+            else:
+                break
+        if current_i >= len(ref_pos_s):
+            return output
+        count += max(delta_r, delta_q)
+        pos_r += delta_r
+        pos_q += delta_q
+        cigar_index += 1
+    while current_i < len(ref_pos_s):
+        output[ref_pos_s[current_i]]=[None, None]
+        current_i +=1
+    return output
+
+class ClusterReads:
+    def __init__(self, bam, variants):
+        self.variants = variants
+        self.chrom = variants[0][0]
+        self.min_pos = variants[0][1]
+        self.max_pos = variants[-1][1]
+        self.reads = []
+        for read_i in bam.fetch(self.chrom, self.min_pos - 1, self.max_pos):
+            if read_i.is_unmapped or not dedup_test(read_i) or read_i.seq is None:
+                continue
+            self.reads.append(read_i)
+
+        done_i = -1
+        n = len(variants)
+        self.var_reads = [[] for i in range(len(variants))]
+        self.read_vars = [[] for i in range(len(self.reads))]
+        for i, read in enumerate(self.reads):
+            for j in range(done_i + 1, n):
+                pos = variants[j][1]
+                if read.reference_start >= pos:
+                    done_i += 1
+                    continue
+                if pos <= read.reference_end:
+                    self.var_reads[j].append(i)
+                    self.read_vars[i].append(j)
+        unused_reads = set(range(len(self.reads)))-set([i for j in self.var_reads for i in j])
+        for i in unused_reads:
+            self.reads[i] = None
+        self.aligned_pairs = []
+        for i, read in enumerate(self.reads):
+            if i not in unused_reads:
+                self.aligned_pairs.append(read.get_aligned_pairs())
+            else:
+                self.aligned_pairs.append(None)
+        self.read_pos_for_ref_pos = []
+        for i, read in enumerate(self.reads):
+            if i not in unused_reads:
+                self.read_pos_for_ref_pos.append(get_read_pos_for_ref_pos(read, 
+                    [self.variants[j][1]-1 for j in self.read_vars[i]]))
+            else:
+                self.read_pos_for_ref_pos.append(None)
+
+
+    def get_alignment_features(self, var_index, ref_base, first_alt, min_mq=1, min_bq=10):
+        '''
+        bam is the opened file handle of bam file
+        my_coordiate is a list or tuple of 0-based (contig, position)
+        '''
+        my_coordinate = self.variants[var_index][0:2]
+        reads = [self.reads[i] for i in self.var_reads[var_index]]
+        aligned_pairs = [self.aligned_pairs[i] for i in self.var_reads[var_index]]
```

Comment:
It may simplify things if you define a class `AugmentedAlignedRead` which subclasses `pysam.AlignmentSegment` and cache `aligned_pairs` right there. And, perhaps, even provide functions like `get_read_pos_for_ref_pos`.